File: src/lib/edge-workers/deployment.ts
import {
	createEdgeWorker,
	listEdgeWorkers,
	updateEdgeWorker,
	validateEdgeWorker,
} from '../api/edge-workers';
import UserError from '../user-error';
import { buildWorker, readPrebuiltWorker, readWorkerSource } from './index';

import type { DiscoveredWorker, EdgeWorker, EdgeWorkerLocation, EdgeWorkerPhase } from './types';
import type { EdgeWorkerWriteInput } from '../api/edge-workers';

export interface EdgeWorkerDeploymentPlanItem {
	action: 'create' | 'update';
	worker: DiscoveredWorker;
	existing: EdgeWorker | null;
	artifact: { wasmPath: string; base64: string; sizeBytes: number };
	validation: 'passed' | 'skipped';
	phases: EdgeWorkerPhase[];
	input: EdgeWorkerWriteInput & { name: string; wasmBinary: string };
	currentLocation: EdgeWorkerLocation | null;
	proposedLocation: EdgeWorkerLocation | null;
	sourceMode: 'store' | 'omit' | 'preserve';
}

export interface EdgeWorkerDeploymentPlanOptions {
	appId: number;
	envId: number;
	projectDir: string;
	workers: readonly DiscoveredWorker[];
	skipBuild: boolean;
	skipValidate: boolean;
	skipSource: boolean;
}

export type EdgeWorkerAppliedCallback = (
	item: EdgeWorkerDeploymentPlanItem,
	result: EdgeWorker
) => void | Promise< void >;

export class DeploymentApplyError extends Error {
	public readonly appliedNames: string[];
	public readonly failedName: string;
	public readonly unappliedNames: string[];

	constructor(
		appliedNames: string[],
		failedName: string,
		unappliedNames: string[],
		cause: unknown
	) {
		super( `Failed to apply edge worker "${ failedName }".`, { cause } );
		this.name = 'DeploymentApplyError';
		this.appliedNames = appliedNames;
		this.failedName = failedName;
		this.unappliedNames = unappliedNames;
	}
}

function prepareArtifact( options: EdgeWorkerDeploymentPlanOptions, worker: DiscoveredWorker ) {
	if ( options.skipBuild ) {
		return readPrebuiltWorker( options.projectDir, worker );
	}
	return buildWorker( options.projectDir, worker );
}

async function prepareValidation(
	options: EdgeWorkerDeploymentPlanOptions,
	worker: DiscoveredWorker,
	artifact: EdgeWorkerDeploymentPlanItem[ 'artifact' ]
): Promise< Pick< EdgeWorkerDeploymentPlanItem, 'validation' | 'phases' > > {
	if ( options.skipValidate ) {
		return { validation: 'skipped', phases: [] };
	}

	const result = await validateEdgeWorker( options.envId, artifact.base64 );
	if ( result.valid !== true ) {
		const errors = result.errors.join( '; ' ) || 'unknown error';
		throw new UserError( `worker "${ worker.manifest.name }" failed validation: ${ errors }` );
	}

	return { validation: 'passed', phases: result.phases };
}

function proposedLocationFor(
	worker: DiscoveredWorker,
	existing: EdgeWorker | null,
	hasLocation: boolean,
	currentLocation: EdgeWorkerLocation | null
): EdgeWorkerLocation | null {
	if ( ! existing ) {
		return worker.manifest.location ?? null;
	}
	if ( ! hasLocation ) {
		return currentLocation;
	}
	return worker.manifest.location ?? null;
}

function prepareInput(
	worker: DiscoveredWorker,
	existing: EdgeWorker | null,
	artifact: EdgeWorkerDeploymentPlanItem[ 'artifact' ],
	source: string | undefined,
	hasLocation: boolean
): EdgeWorkerDeploymentPlanItem[ 'input' ] {
	const input: EdgeWorkerDeploymentPlanItem[ 'input' ] = {
		name: worker.manifest.name,
		wasmBinary: artifact.base64,
	};
	if ( worker.manifest.on_failure ) {
		input.onFailure = worker.manifest.on_failure;
	}
	if ( source ) {
		input.source = source;
	}
	if ( existing && hasLocation ) {
		input.location = worker.manifest.location ?? null;
	} else if ( ! existing && worker.manifest.location ) {
		input.location = worker.manifest.location;
	}
	return input;
}

function sourceModeFor(
	skipSource: boolean,
	existing: EdgeWorker | null
): EdgeWorkerDeploymentPlanItem[ 'sourceMode' ] {
	if ( ! skipSource ) {
		return 'store';
	}
	return existing ? 'preserve' : 'omit';
}

async function preparePlanItem(
	options: EdgeWorkerDeploymentPlanOptions,
	worker: DiscoveredWorker,
	existing: EdgeWorker | null
): Promise< EdgeWorkerDeploymentPlanItem > {
	const artifact = prepareArtifact( options, worker );
	const { validation, phases } = await prepareValidation( options, worker, artifact );
	const source = options.skipSource ? undefined : readWorkerSource( worker );
	const hasLocation = Object.hasOwn( worker.manifest, 'location' );
	const currentLocation = existing?.location ?? null;

	return {
		action: existing ? 'update' : 'create',
		worker,
		existing,
		artifact,
		validation,
		phases,
		input: prepareInput( worker, existing, artifact, source, hasLocation ),
		currentLocation,
		proposedLocation: proposedLocationFor( worker, existing, hasLocation, currentLocation ),
		sourceMode: sourceModeFor( options.skipSource, existing ),
	};
}

export async function prepareEdgeWorkerDeploymentPlan(
	options: EdgeWorkerDeploymentPlanOptions
): Promise< EdgeWorkerDeploymentPlanItem[] > {
	const remoteWorkers = await listEdgeWorkers( options.appId, options.envId );
	const remoteWorkersByName = new Map( remoteWorkers.map( worker => [ worker.name, worker ] ) );
	const items: EdgeWorkerDeploymentPlanItem[] = [];

	for ( const worker of options.workers ) {
		const existing = remoteWorkersByName.get( worker.manifest.name ) ?? null;
		// eslint-disable-next-line no-await-in-loop
		items.push( await preparePlanItem( options, worker, existing ) );
	}

	return items;
}

function formatLocation( location: EdgeWorkerLocation | null ): string {
	return location ? `${ location.operator } "${ location.value }"` : 'all requests';
}

export function deploymentPlanRows(
	items: readonly EdgeWorkerDeploymentPlanItem[]
): Record< string, string >[] {
	return items.map( item => ( {
		worker: item.worker.manifest.name,
		action: item.action,
		active: item.existing?.active ? 'yes' : 'no',
		current_scope: formatLocation( item.currentLocation ),
		proposed_scope: formatLocation( item.proposedLocation ),
		validation: item.validation,
		phases: item.phases.join( ', ' ) || 'none',
		bytes: String( item.artifact.sizeBytes ),
		source: item.sourceMode,
	} ) );
}

async function applyPlanItem(
	envId: number,
	item: EdgeWorkerDeploymentPlanItem
): Promise< EdgeWorker > {
	if ( item.action === 'create' ) {
		return createEdgeWorker( envId, item.input );
	}
	if ( ! item.existing ) {
		throw new Error( `Update plan for "${ item.worker.manifest.name }" has no existing worker.` );
	}
	return updateEdgeWorker( envId, item.existing.id, item.input );
}

export async function applyEdgeWorkerDeploymentPlan(
	envId: number,
	items: readonly EdgeWorkerDeploymentPlanItem[],
	onApplied: EdgeWorkerAppliedCallback
): Promise< void > {
	const appliedNames: string[] = [];

	for ( const [ index, item ] of items.entries() ) {
		const name = item.worker.manifest.name;
		let result: EdgeWorker;
		try {
			// eslint-disable-next-line no-await-in-loop
			result = await applyPlanItem( envId, item );
		} catch ( cause ) {
			throw new DeploymentApplyError(
				[ ...appliedNames ],
				name,
				items.slice( index + 1 ).map( remaining => remaining.worker.manifest.name ),
				cause
			);
		}

		appliedNames.push( name );
		// eslint-disable-next-line no-await-in-loop
		await onApplied( item, result );
	}
}

File: internal/edgeworkers/deployment.go
package edgeworkers

import (
	"context"
	"errors"
	"fmt"
	"strings"
)

type ArtifactBuilder interface {
	Build(context.Context, string, LocalWorker) (Artifact, error)
}
type Service struct {
	API     API
	Builder ArtifactBuilder
}
type PlanOptions struct {
	AppID, EnvID                                int64
	ProjectDir                                  string
	Workers                                     []LocalWorker
	SkipBuild, SkipValidate, SkipSource, Enable bool
}
type PlanItem struct {
	Action                            string
	Worker                            LocalWorker
	Existing                          *Worker
	Artifact                          Artifact
	Validation                        string
	Phases                            []string
	Input                             WriteInput
	CurrentLocation, ProposedLocation *Location
	SourceMode                        string
	EnableAfterDeploy, IntendedActive bool
}
type ApplyError struct {
	AppliedNames      []string
	FailedName        string
	UnappliedNames    []string
	Cause             error
	Stage             string
	UploadCompleted   bool
	ActiveAfterUpload *bool
}

func (e *ApplyError) Error() string {
	if e.Cause != nil {
		return e.Cause.Error()
	}
	return "Deployment failed."
}
func (e *ApplyError) Unwrap() error { return e.Cause }

func (s Service) PreparePlan(ctx context.Context, opts PlanOptions) ([]PlanItem, error) {
	remote, err := s.API.List(ctx, opts.AppID, opts.EnvID)
	if err != nil {
		return nil, err
	}
	byName := map[string]*Worker{}
	for i := range remote {
		byName[remote[i].Name] = &remote[i]
	}
	items := make([]PlanItem, 0, len(opts.Workers))
	for _, w := range opts.Workers {
		item := PlanItem{Action: "create", Worker: w, Existing: byName[w.Manifest.Name], Validation: "skipped", Phases: []string{}, SourceMode: "store", EnableAfterDeploy: opts.Enable, IntendedActive: opts.Enable}
		if opts.SkipBuild {
			item.Artifact, err = ReadPrebuilt(opts.ProjectDir, w)
		} else if s.Builder != nil {
			item.Artifact, err = s.Builder.Build(ctx, opts.ProjectDir, w)
		} else {
			return nil, errors.New("No worker compiler configured.")
		}
		if err != nil {
			return nil, err
		}
		if !opts.SkipValidate {
			result, e := s.API.Validate(ctx, opts.EnvID, item.Artifact.Base64)
			if e != nil {
				return nil, e
			}
			if !result.Valid {
				details := strings.Join(result.Errors, "; ")
				if details == "" {
					details = "unknown error"
				}
				return nil, fmt.Errorf("worker \"%s\" failed validation: %s", w.Manifest.Name, details)
			}
			item.Validation = "passed"
			item.Phases = result.Phases
		}
		item.Input = WriteInput{Name: w.Manifest.Name, WASMBinary: item.Artifact.Base64, OnFailure: w.Manifest.OnFailure, Location: w.Manifest.Location}
		if !opts.SkipSource {
			source, e := ReadWorkerSource(w)
			if e != nil {
				return nil, e
			}
			item.Input.Source = &source
		} else {
			item.SourceMode = "omit"
		}
		item.ProposedLocation = w.Manifest.Location.Value
		if item.Existing != nil {
			item.Action = "update"
			item.CurrentLocation = item.Existing.Location
			item.IntendedActive = opts.Enable || item.Existing.Active
			if !w.Manifest.Location.Present {
				item.ProposedLocation = item.CurrentLocation
			}
			if opts.SkipSource {
				item.SourceMode = "preserve"
			}
		} else if item.Input.Location.Value == nil {
			item.Input.Location = LocationValue{}
		}
		items = append(items, item)
	}
	return items, nil
}

func newApplyError(items []PlanItem, index int, applied []string, stage string, uploaded bool, active *bool, cause error) *ApplyError {
	e := &ApplyError{AppliedNames: append([]string{}, applied...), FailedName: items[index].Worker.Manifest.Name, UnappliedNames: []string{}, Stage: stage, UploadCompleted: uploaded, ActiveAfterUpload: active, Cause: cause}
	for _, item := range items[index+1:] {
		e.UnappliedNames = append(e.UnappliedNames, item.Worker.Manifest.Name)
	}
	return e
}
func (s Service) ApplyPlan(ctx context.Context, envID int64, items []PlanItem, onApplied func(PlanItem, Worker) error) error {
	applied := []string{}
	for i, item := range items {
		var result Worker
		var err error
		if item.Action == "create" {
			result, err = s.API.Create(ctx, envID, item.Input)
		} else if item.Existing == nil {
			err = fmt.Errorf("Update plan for \"%s\" has no existing worker.", item.Worker.Manifest.Name)
		} else {
			result, err = s.API.Update(ctx, envID, item.Existing.ID, item.Input)
		}
		if err != nil {
			return newApplyError(items, i, applied, "upload", false, nil, err)
		}
		if item.EnableAfterDeploy && !result.Active {
			active := result.Active
			result, err = s.API.SetActive(ctx, envID, result.ID, true)
			if err != nil {
				return newApplyError(items, i, applied, "enable", true, &active, err)
			}
		}
		applied = append(applied, item.Worker.Manifest.Name)
		if onApplied != nil {
			if err := onApplied(item, result); err != nil {
				return err
			}
		}
	}
	return nil
}
